File: 2780-minimum-index-of-a-valid-split/2780-minimum-index-of-a-valid-split.py
class Solution:
    def minimumIndex(self, nums: List[int]) -> int:
        cnt = 0
        domi_num = 0
        for num in nums:
            if cnt == 0:
                domi_num = num
            if domi_num == num:
                cnt += 1
            else:
                cnt -= 1
        n = len(nums)
        tot = nums.count(domi_num)
        if n == tot * 2 - 1:
            return -1
        cnt = 0
        for i, num in enumerate(nums):
            if num == domi_num:
                cnt += 1
            if cnt * 2 > i + 1 and (tot - cnt) * 2 > n - i - 1:
                return i
        return -1

File: 2780-minimum-index-of-a-valid-split/2780-minimum-index-of-a-valid-split.cpp
class Solution {
    int mooresVotingAlgo(vector<int>& nums){
        int majority_element = nums[0];
        int freq = 1;
        for(int i=1;i<nums.size();++i){
            if(nums[i]!=majority_element)   freq--;
            else                            freq++;
            
            if(freq==0){
                majority_element = nums[i];
                freq = 1;
            }
        }
        return majority_element;
    }
public:
    int minimumIndex(vector<int>& nums) {
        int n=nums.size();
        //Step-1: Find Majority element
        int majority_element = mooresVotingAlgo(nums);

        //Step-2: Count frequency of majority_element
        int max_freq = 0;
        for(int i=0;i<n;++i)
            if(nums[i]==majority_element)
                max_freq++;

        //Step-3: Now find minimum valid split
        int prefix_count = 0;
        for(int i=0;i<n-1;++i){
            if(nums[i]==majority_element){
                prefix_count++;
                max_freq--;
            }
            if((prefix_count > (i+1)/2) and (max_freq > (n-i-1)/2))
                return i;
        }
        return -1;
    }
};